src/vsr/replica_format.zig

```diff
@@ -26,195 +22,287 @@ pub fn format(
                     write_sectors_callback,
-                    &self.write,
+                    &self.writes[self.writes_pending],
                     padding_buffer,
                     .grid_padding,
                     0,
                 );
-                self.formatting = true;
-                while (self.formatting) storage.run();
+                self.writes_pending += 1;
             }
         }
 
-        fn write_sectors_callback(write: *Storage.Write) void {
-            const self: *ReplicaFormat = @alignCast(@fieldParentPtr("write", write));
+        fn format_and_tick(
+            self: *ReplicaFormat,
+            storage: *Storage,
+            superblock: *SuperBlock,
+            superblock_options: SuperBlock.FormatOptions,
+        ) void {
+            assert(formating_instance == null);
+            formating_instance = self;
+
+            assert(self.writes_pending == self.writes_expected);
+
+            superblock.format(
+                format_superblock_callback,
+                &self.superblock_context,
+                superblock_options,
+            );
+
+            self.formatting = true;
+            self.formatting_superblock = true;
+            while (self.formatting or self.formatting_superblock) storage.run();
+
+            formating_instance = null;
+        }
+
+        fn write_sectors_callback(_: *Storage.Write) void {
+            const self: *ReplicaFormat = formating_instance.?;
             assert(self.formatting);
-            self.formatting = false;
+
+            self.writes_pending -= 1;
+
+            if (self.writes_pending == 0) {
+                self.formatting = false;
+            }
         }
 
         fn format_superblock_callback(superblock_context: *SuperBlock.Context) void {
             const self: *ReplicaFormat =
                 @alignCast(@fieldParentPtr("superblock_context", superblock_context));
-            assert(self.formatting);
-            self.formatting = false;
+            assert(self.formatting_superblock);
+            self.formatting_superblock = false;
         }
     };
 }
 
+/// Write a WAL header into the start of header_buffer, and return a pointer to the written header
+/// as well as the offset its matching prepare would go into.
+pub fn wal_header(cluster: u128, slot: u64, header_buffer: []u8) struct { *Header.Prepare, u64 } {
+    assert(slot < constants.journal_slot_count);
+    assert(slot * @sizeOf(Header.Prepare) < constants.journal_size_headers);
+    assert(@sizeOf(Header.Prepare) < constants.sector_size);
+    assert(header_buffer.len >= @sizeOf(Header.Prepare));
+
+    stdx.maybe(header_buffer.len % constants.sector_size == 0);
+
+    const prepare_offset = slot * constants.message_size_max;
+    assert(prepare_offset <= constants.journal_size_prepares);
+    assert(prepare_offset % @sizeOf(Header) == 0);
+    assert(prepare_offset % constants.sector_size == 0);
+
+    const header: *Header.Prepare = @alignCast(
```

Comment:
If we keep this function, we should push alignment requirement into the signature